src/v/cloud_storage_clients/client_pool.cc

```diff
@@ -272,7 +272,9 @@ client_pool::acquire(ss::abort_source& as) {
                 // client connections then wait util one of the clients is
                 // freed.
                 co_await ssx::with_timeout_abortable(
-                  _cvar.wait(), model::no_timeout, as);
+                  _cvar.wait([this] { return !_pool.empty(); }),
```

Comment:
I don't think that the `condition_variable` in seastar has spurious wakeups. Normal condition variable uses spurious wakups as an optimization to avoid storing and synchronizing all waiters all the time. In seastar it stores an intrusive list of all waiters implicitly so there are no unexpected wakeups at all.

It has another problem. The wait is async so when the callback is scheduled the condition might be true. But when the `wait` method exits the condition might already be false. In this case the log line below may say that the pool size is 0 because between the scheduling of the callback and scheduling of the continuation that follows `wait` something else removed element from the pool.

Reply:
so maybe the description of the commit is incorrect or the problem is different

Reply:
You are most likely correct. The error was caused by a bug in the test rather than here. Replaced this commit with one fixing the test instead.